### PR TITLE
add only valid json option when parsing json.

### LINKED ISSUE #121: `parse-string` parses invalid JSON
```
(parse-string "\"hello\" : 123}")
;;=> "hello"

(parse-string-strict "\"hello\" : 123}")
;;=> "hello"
```

### PR DESCRIPTION
**Pull request by lucacervello**
Hi,

this is my attempt to fix #121 #79 #65 #52,

### CODE REVIEW

**Comment by dakrone** [comment]
@lucacervello hmm.. I am concerned that this approach incurs the cost for the check for all parsing.

How would you feel about adding a new parser in `core` that does the check, and the other parsers can continue to ignore it? Maybe `parse-exact`?

**Comment by lucacervello** [comment]
@dakrone I understand your concern and i'm ok with adding `parse-exact` to `core`.

**Comment by dakrone** [comment]
@lucacervello great, is that something you'd like to do as part of this PR?

**Comment by lucacervello** [comment]
@dakrone yes definitely.

**Comment by lucacervello** [comment]
@dakrone is only one `parse-exact` enough ? or we want:
- `parse-string` and `parse-string-exact`
- `parse-string-strict` and `parse-string-strict-exact`
- `parse-stream` and `parse-stream-exact`
- `parse-smile` and `parse-smile-exact`

If we want an `exact` function for every `parse`
Maybe a macro `with-exact` or `exact` that wrap that parsing functions is a better solution.
Something like this
``` clojure
(with-exact
  (parse-string "{\"foo\": 1}"))
```
I think `with-exact` can works but i'm not a macro expert 😄

**Comment by dakrone** [comment]
I think that would work, as long as the macro simply added the validation after the `parse-string` function instead of modifying it directly. Want to give it a shot?

**Comment by lucacervello** [comment]
Yes, sure.

**Comment by lucacervello** [comment]
@dakrone i think the "macro approach" can't work. I need to access the JsonParser object and i can't from outside.
 Maybe an `exact` function for every `parse` is a better idea.

**Comment by dakrone** [comment]
Okay, I'm okay either way. If we *do* have to duplicate each, maybe we can just put them in a different namespace and name them the same thing, rather than add more things to the core one?

Maybe `cheshire.exact`? What do you think?

**Comment by lucacervello** [comment]
I'm okay with it. I'll give it a try.

**Comment by lucacervello** [comment]
@dakrone I've done only `parse-string` and `parse-string-strict`. 
I don't know if translate `parse-stream` and `parse-smile` make any sense.
There are cases where the stream isn't a valid json ?

File: src/cheshire/exact.clj [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  (:import (java.io StringReader Reader BufferedReader
+                    Writer)
+           (com.fasterxml.jackson.core JsonFactory)))
+
+(defn- exact-parse [jp parsed]
+  (let [valid-json? (try (nil? (.nextToken jp))
+                         (catch Exception _ false))]
+    (if valid-json?
+      parsed
+      (throw (Exception. "Invalid json")))))
```
Comment (dakrone) [inline src/cheshire/exact.clj]: Can this be `IllegalArgumentException` and the message be `Invalid JSON, expected exactly one parseable object but multiple objects were found"

**Comment by dakrone** [comment]
Merged this, thanks so much @lucacervello! I'll release 5.8.0 with this shortly

**Pull request merged**